common/types/offer.go

```diff
@@ -32,6 +38,7 @@ func HexToHash(s string) (Hash, error) {
 // Offer represents a swap offer
 type Offer struct {
 	ID            Hash
+	idMu          sync.Mutex `json:"-"` // prevents race conditions on the first call to GetID
```

Comment:
In a previous edition of this PR, @noot wrote:

> I think I'd rather just have a NewOffer function that must be used that sets the ID. however to enforce that NewOffer needs to be used we're have to un-export the struct values, which would cause json.Marshal not to work, but we could just concatenate the values inside the struct. or we could leave the values exported and make sure we use NewOffer.

I'll make that change before we merge. As part of that change, I was wondering if we should do one of:

1. Not use the serialized data from the offer when generating the offer ID. Since the hash includes 8 bytes of random data that we don't store, there is no way for us to do any checks based on the serialized data used for the hash.
2. Keep the serialized data, but also store the random bytes as a nonce. If the nonce is stored, we could do checks on the offer values relative to the ID if there is a use case for doing this.